enkit/outputs/commands.go

```diff
@@ -0,0 +1,187 @@
+}
+
+type Unmount struct {
+	*cobra.Command
+	root *Root
+}
+
+func NewUnmount(root *Root) *Unmount {
+	command := &Unmount{
+		Command: &cobra.Command{
+			Use:   "unmount [invocation ID]",
+			Short: "Unmount the build outputs of a particular invocation",
+			Example: `  $ enkit outputs unmount 73d4a9f0-a0c4-4cb2-80eb-b4b4b9720d07
+	Unmounts outputs from build 73d4a9f0-a0c4-4cb2-80eb-b4b4b9720d07 from the
+	default location.`,
+			Aliases: []string{"umount"},
+			Args:    cobra.ExactArgs(1),
+		},
+		root: root,
+	}
+	command.Command.RunE = command.Run
+	return command
+}
+
+func (c *Unmount) Run(cmd *cobra.Command, args []string) error {
+	return fmt.Errorf("`enkit outputs unmount` is unimplemented")
+}
+
+type Run struct {
+	*cobra.Command
+	root *Root
+
+	InvocationID string
+	DirPath      string
+	ZipPath      string
+}
+
+func NewRun(root *Root) *Run {
+	command := &Run{
+		Command: &cobra.Command{
+			Use:   "run",
+			Short: "Mount build artifacts and execute an optional command on them",
+			Example: `  $ enkit outputs run --invocation=73d4a9f0-a0c4-4cb2-80eb-b4b4b9720d07
+	Launches a shell with build outputs from a particular build re-rooted so that
+	paths are correct within the outputs themselves.
+
+  $ enkit outputs run --dir=/tmp/some_dir
+	Launches a shell with build outputs in /tmp/some_dir re-rooted so that paths
+	are correct within the outputs themselves.
+
+  $ enkit outputs run --zip=/tmp/some.zip
+	Unpacks the named zip into a tempdir, and then reroots the artifacts so that
+	paths are correct within the outputs themselves.
+
+  $ enkit outputs run --zip=/tmp/some-zip -- find .
+	Runs "find ." in the unpacked zip after it is rerooted.`,
+		},
+		root: root,
+	}
+
+	command.Command.RunE = command.Run
+	command.Command.PreRunE = command.validate
+	command.Flags().StringVar(&command.InvocationID, "invocation_id", "", "If set, the build's invocation ID from which to mount artifacts")
+	command.Flags().StringVar(&command.DirPath, "dir_path", "", "If set, the path to already-unpacked artifacts to reroot")
+	command.Flags().StringVar(&command.ZipPath, "zip_path", "", "If set, the path to a zipped outputs file to unpack and re-root")
+	return command
+}
+
+func (c *Run) Run(cmd *cobra.Command, args []string) error {
+	return fmt.Errorf("`enkit outputs run` is unimplemented")
+}
+
+func (c *Run) validate(cmd *cobra.Command, args []string) error {
+	setCount := 0
+	for _, f := range []string{c.InvocationID, c.DirPath, c.ZipPath} {
+		if f != "" {
+			setCount++
+		}
+	}
+	switch {
+	case setCount == 0:
+		return fmt.Errorf("One of --invocation_id, --dir_path, --zip_path must be set")
+	case setCount >= 2:
+		return fmt.Errorf("Only one of --invocation_id, --dir_path, --zip_path may be set")
+	}
+	return nil
+}
+
+type Shutdown struct {
+	*cobra.Command
+	root *Root
+}
+
+func NewShutdown(root *Root) *Shutdown {
+	command := &Shutdown{
+		Command: &cobra.Command{
+			Use:   "shutdown",
```

Comment:
shutdown -> are we using a single mount? or multiple mounts? also, "deletes all links"...

In the document I was proposing using "scratch/" for *hard links* that are hard for tools to distinguish. Is that still the plan?

I was proposing using a bb_clientd mount per invocation as I don't know if scratch/ allows to create directories. If it does, than a single bb_clientd mount is feasible. If it doesn't, then we must have a different bb_clientd mount per invocation id.

Reply:
@aaahrens says you can `mkdir` in the scratch directory, so we get by with one bb_clientd instance per `output_root` directory. `shutdown` will operate on one `output_root` by unmounting bb_clientd (which removes the scratch dir and thus all the hardlinks) and deleting the convenience symlinks within

Reply:
sgtm, then! and I assume the mounting is done dynamically, no need for shutdown.